src/refresh.cc

```diff
@@ -336,6 +336,9 @@ refreshCheck(const StoreEntry * entry, HttpRequest * request, time_t delta)
      */
     const bool revalidateAlways = EBIT_TEST(entry->flags, ENTRY_REVALIDATE_ALWAYS);
     if (revalidateAlways || (staleness > -1 &&
+#if USE_HTTP_VIOLATIONS
+                             !R->flags.ignore_must_revalidate &&
+#endif
                              EBIT_TEST(entry->flags, ENTRY_REVALIDATE_STALE))) {
```

Comment:
This change has two problems AFAICT:

* It applies ignore-must-revalidate configuration option to responses with `Cache-Control: max-age`. That is wrong.
* It applies ignore-must-revalidate configuration option to responses cached _before_ the option was enabled. The proposed documentation does not make it clear that the option affects already cached responses. We should be explicit about this unexpected (for some) "backdating" effect or unexpected (for some) lack of thereof.

The fix(es) would depend on whether we want ignore-must-revalidate to apply to previously cached responses. AFAICT, if we want the simplest solution, then we should answer "no" and

* "move" this check, in its REFRESH_OVERRIDE() form, to where hasMustRevalidate() is called by HttpStateData::haveParsedReplyHeaders(), eventually resulting in setting ENTRY_REVALIDATE_STALE;
* adjust proposed documentation to explicitly exclude already cached responses from the configuration option scope/control.

Reply:
Thanks for the review!

Maybe I'm missing something, but doesn't `Cache-Control: max-age` get processed [after this](https://github.com/squid-cache/squid/blob/master/src/refresh.cc#L406) (checked independently)?

It seems to me like the easiest thing would be to:
1. Document that this option does not affect `Cache-Control: proxy-revalidate`, presumably someone who has turned on that option knows what they are doing (and if they don't, a separate option probably makes sense)
3. Allow the option to apply to cached information from before the option was enabled, to match the original behavior (to minimize unexpected changes, as mentioned in your other comment)
4. Document this behavior so that it is clear that the option is about _revalidation_ of cached objects, not really the original caching

Proposed documentation:
> ignore-must-revalidate ignores any ``Cache-Control: must-revalidate`` headers received from a server. This means that previously cached content (potentially cached before the option was enabled) will be reused instead of checking whether the server has updated content. Doing this VIOLATES the HTTP standard. Enabling this feature could make you liable for problems which it causes. Please note this option does not change the behavior of ``Cache-Control: proxy-revalidate``, but does take affect for authenticated transactions where ``Cache-Control: must-revalidate`` or ``Cache-Control: no-cache`` is set (without parameters) for backwards-compatibility with the pre-v4 behavior.

Note: The authenticated transaction part is not relevant in my use case, but I figured in re-introducing this option it would be best to have it behave the same way it did before.

Reply:
FYI: this is a generic check that the object is stale and therefor needs revalidating. If it gets removed the objects will likely go through a full re-fetch which is counter to intentions.
 Its relationship to ```must-revalidate``` is through how the revalidate requirement implies immediate staleness. If you can identify the code converting response ```Cache-Control:must-revalidate``` into the ```staleness=0``` value that would likely be a better place to put the override check.

I would prefer splitting this complicated if-condition up so the effects of the relevant cache-control values can be clearly seen and this ignore/override can be done to them individually. Though that may require expanding out the ENTRY_REVALIDATE_STALE flag meanings.

Reply:
> Alex: Sorry, I should have said `s-maxage` not `max-age`...

Hm... I wrote that [comment](https://github.com/squid-cache/squid/pull/1196#discussion_r1038519601) a few days ago, but only received an email notification about it now, after adding a couple of fresh comments to other change requests. It looks like my comment was stuck for all those days (in a pending state?). GitHub still shows the correct (from my point of view) comment order in this change request, but my comment has a much later timestamp than the subsequent Amos' comment. Anyway, I think that stuck/resurfaced comment is still valid. Sorry about the mess!